# Segmenting and Clustering Neighborhoods in Toronto

All three parts of this week are in this notebook only, kindly scroll down for part 2 and 3 <br>
Suraj Tiwari

In [1]:
!conda install -c conda-forge beautifulsoup4 --yes

!conda install -c conda-forge geopy --yes

!conda install -c conda-forge folium=0.5.0 --yes

print('Libraries installed!')

Fetching package metadata .............
Solving package specifications: .

Package plan for installation in environment /opt/conda/envs/DSX-Python35:

The following packages will be UPDATED:

    beautifulsoup4: 4.6.0-py35h442a8c9_1 --> 4.6.3-py35_0 conda-forge

beautifulsoup4 100% |################################| Time: 0:00:00   1.45 MB/s
Fetching package metadata .............
Solving package specifications: .

Package plan for installation in environment /opt/conda/envs/DSX-Python35:

The following NEW packages will be INSTALLED:

    geographiclib: 1.49-py_0   conda-forge
    geopy:         1.17.0-py_0 conda-forge

geographiclib- 100% |################################| Time: 0:00:00   1.05 MB/s
geopy-1.17.0-p 100% |################################| Time: 0:00:00   1.67 MB/s
Fetching package metadata .............
Solving package specifications: .

Package plan for installation in environment /opt/conda/envs/DSX-Python35:

The following NEW packages will be INSTALLED:

    altair:

In [2]:
import numpy as np
import pandas as pd
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import requests
from pandas.io.json import json_normalize
import json

import requests

from bs4 import BeautifulSoup

from geopy.geocoders import Nominatim

import folium
import matplotlib.cm as cm
import matplotlib.colors as colors

from sklearn.cluster import KMeans

print('Libraries imported!')

Libraries imported!


##  Web scraping for Toronto neighborhood and build a clean dataframe

Use the Notebook to build the code to scrape the following Wikipedia page, https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M, in order to obtain the data that is in the table of postal codes and to transform the data into a pandas dataframe.I have used BeautifulSoup for the task.

In [3]:
# open Wiki page with Beautiful Soup
data = requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').text
soup = BeautifulSoup(data, 'html.parser')

In [4]:
postalCodeList = []
boroughList = []
neighborhoodList = []

for row in soup.find('table').find_all('tr'):
    cells = row.find_all('td')
    if(len(cells) > 0):
        postalCodeList.append(cells[0].text)
        boroughList.append(cells[1].text)
        neighborhoodList.append(cells[2].text.rstrip('\n')) # remove the new line char from neighborhood cell

In [5]:
toronto_neighorhood = [('PostalCode', postalCodeList),
                      ('Borough', boroughList),
                      ('Neighborhood', neighborhoodList)]
toronto_df = pd.DataFrame.from_dict(dict(toronto_neighorhood))
toronto_df.head()

Borough      Neighborhood PostalCode
0      Not assigned      Not assigned        M1A
1      Not assigned      Not assigned        M2A
2        North York         Parkwoods        M3A
3        North York  Victoria Village        M4A
4  Downtown Toronto      Harbourfront        M5A


### We will remove the rows where postal codes do not belongs to any borough.  


In [6]:
toronto_df_dropna = toronto_df[toronto_df.Borough != 'Not assigned'].reset_index(drop=True)
toronto_df_dropna.head()

Borough      Neighborhood PostalCode
0        North York         Parkwoods        M3A
1        North York  Victoria Village        M4A
2  Downtown Toronto      Harbourfront        M5A
3  Downtown Toronto       Regent Park        M5A
4        North York  Lawrence Heights        M6A

In [7]:
toronto_df_grouped = toronto_df_dropna.groupby(['PostalCode','Borough'], as_index=False).agg(lambda x: ','.join(x))
toronto_df_grouped.head()

PostalCode      Borough                          Neighborhood
0        M1B  Scarborough                         Rouge,Malvern
1        M1C  Scarborough  Highland Creek,Rouge Hill,Port Union
2        M1E  Scarborough       Guildwood,Morningside,West Hill
3        M1G  Scarborough                                Woburn
4        M1H  Scarborough                             Cedarbrae

We will replace the 'Not assigned' with the value of the M7A Queen's Park

In [8]:
na_neigh_rows = toronto_df_grouped.Neighborhood == 'Not assigned'
toronto_df_grouped.loc[na_neigh_rows, 'Neighborhood'] = toronto_df_grouped.loc[na_neigh_rows, 'Borough']
toronto_df_grouped[na_neigh_rows]

PostalCode       Borough  Neighborhood
85        M7A  Queen's Park  Queen's Park

In [9]:
toronto_df_cleaned = toronto_df_grouped
toronto_df_cleaned.shape

(103, 3)

The final dataframe has 103 rows

## Part 2: Getting coordinates and add to the Toronto DataFrame  

The csv file format has:
- 3 columns: Postal Code, Latitude and Longitude
- 103 rows: corresponding to 103 postal codes in our toronto dataframe

In [10]:
!wget -q -O "toronto_coordinates.csv" http://cocl.us/Geospatial_data
print('Coordinates downloaded!')
coors = pd.read_csv('toronto_coordinates.csv')

Coordinates downloaded!


In [11]:
print(coors.shape)
coors.head()

(103, 3)


Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

We will merge the two dataframes.  
To get the correct result independant on the order of data, we need to set indexes of two dataframes to its Postal Code columns

In [12]:
toronto_df_temp = toronto_df_cleaned.set_index('PostalCode')
coors_temp = coors.set_index('Postal Code')
toronto_df_coors = pd.concat([toronto_df_temp, coors_temp], axis=1, join='inner')

Reset index and we will get the toronto dataframe with coordinates

In [13]:
toronto_df_coors.index.name = 'PostalCode'
toronto_df_coors.reset_index(inplace=True)

Check the toronto dataframe with coordinates

In [14]:
print(toronto_df_coors.shape)
toronto_df_coors.head()

(103, 5)


PostalCode      Borough                          Neighborhood   Latitude  \
0        M1B  Scarborough                         Rouge,Malvern  43.806686   
1        M1C  Scarborough  Highland Creek,Rouge Hill,Port Union  43.784535   
2        M1E  Scarborough       Guildwood,Morningside,West Hill  43.763573   
3        M1G  Scarborough                                Woburn  43.770992   
4        M1H  Scarborough                             Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

## Part 3: Explore and cluster the neighborhoods in Toronto

In [15]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="tl-toronto-neigh")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinates of Toronto are 43.653963, -79.387207.


**Using Folium to create a Map of Toronto with Boroughs markers on top**

In [16]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

for lat, long, post, borough, neigh in zip(toronto_df_coors['Latitude'], toronto_df_coors['Longitude'], toronto_df_coors['PostalCode'], toronto_df_coors['Borough'], toronto_df_coors['Neighborhood']):
    label = "{} ({}): {}".format(borough, post, neigh)
    popup = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, long],
        radius=5,
        popup=popup,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)
    
map_toronto

**Reduce the number of Boroughs to explore**
To reduce the numbers of calls to FourSquare API, we will only explore boroughs that have Toronto in their names.

In [17]:
toronto_boroughs = ['East Toronto', 'Central Toronto', 'Downtown Toronto', 'West Toronto']
toronto_central_df = toronto_df_coors[toronto_df_coors['Borough'].isin(toronto_boroughs)].reset_index(drop=True)
print(toronto_central_df.shape)
toronto_central_df.head()

(38, 5)


PostalCode          Borough                   Neighborhood   Latitude  \
0        M4E     East Toronto                    The Beaches  43.676357   
1        M4K     East Toronto    The Danforth West,Riverdale  43.679557   
2        M4L     East Toronto  The Beaches West,India Bazaar  43.668999   
3        M4M     East Toronto                Studio District  43.659526   
4        M4N  Central Toronto                  Lawrence Park  43.728020   

   Longitude  
0 -79.293031  
1 -79.352188  
2 -79.315572  
3 -79.340923  
4 -79.388790

In [18]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=12)

for lat, long, post, borough, neigh in zip(toronto_central_df['Latitude'], toronto_central_df['Longitude'], toronto_central_df['PostalCode'], toronto_central_df['Borough'], toronto_central_df['Neighborhood']):
    label = "{} ({}): {}".format(borough, post, neigh)
    popup = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, long],
        radius=5,
        popup=popup,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)
    
map_toronto

**Using FourSquare API to explore the Boroughs**

In [21]:
# The code was removed by Watson Studio for sharing.

Get recommend places inside or near each Borough in Toronto Central

In [22]:
radius = 500
LIMIT = 100

venues = []

for lat, long, post, borough, neighborhood in zip(toronto_central_df['Latitude'], toronto_central_df['Longitude'], toronto_central_df['PostalCode'], toronto_central_df['Borough'], toronto_central_df['Neighborhood']):
    url = "https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}".format(
        CLIENT_ID,
        CLIENT_SECRET,
        VERSION,
        lat,
        long,
        radius, 
        LIMIT)
    
    results = requests.get(url).json()["response"]['groups'][0]['items']
    
    for venue in results:
        venues.append((
            post, 
            borough,
            neighborhood,
            lat, 
            long, 
            venue['venue']['name'], 
            venue['venue']['location']['lat'], 
            venue['venue']['location']['lng'],  
            venue['venue']['categories'][0]['name']))

In [23]:
venues_df = pd.DataFrame(venues)
venues_df.columns = ['PostalCode', 'Borough', 'Neighborhood', 'BoroughLatitude', 'BoroughLongitude', 'VenueName', 'VenueLatitude', 'VenueLongitude', 'VenueCategory']
print(venues_df.shape)
venues_df.head()

(1704, 9)


PostalCode       Borough Neighborhood  BoroughLatitude  BoroughLongitude  \
0        M4E  East Toronto  The Beaches        43.676357        -79.293031   
1        M4E  East Toronto  The Beaches        43.676357        -79.293031   
2        M4E  East Toronto  The Beaches        43.676357        -79.293031   
3        M4E  East Toronto  The Beaches        43.676357        -79.293031   
4        M4E  East Toronto  The Beaches        43.676357        -79.293031   

             VenueName  VenueLatitude  VenueLongitude         VenueCategory  
0            Starbucks      43.678798      -79.298045           Coffee Shop  
1  Grover Pub and Grub      43.679181      -79.297215                   Pub  
2    Glen Manor Ravine      43.676821      -79.293942                 Trail  
3        Upper Beaches      43.680563      -79.292869          Neighborhood  
4      Beaches Fitness      43.680319      -79.290991  Gym / Fitness Center

**Check how many venues returned for each postal code**

In [24]:
venues_df.groupby(['PostalCode', 'Borough', 'Neighborhood'])['VenueName'].count()

PostalCode  Borough           Neighborhood                                                                                        
M4E         East Toronto      The Beaches                                                                                               5
M4K         East Toronto      The Danforth West,Riverdale                                                                              42
M4L         East Toronto      The Beaches West,India Bazaar                                                                            20
M4M         East Toronto      Studio District                                                                                          40
M4N         Central Toronto   Lawrence Park                                                                                             4
M4P         Central Toronto   Davisville North                                                                                          7
M4R         Central Toronto   North Toron

**Check how many kinds of venues categories is returned**

In [25]:
len(venues_df['VenueCategory'].unique())

233

**Analyze venues in each area**

In [26]:
# one hot encoding
toronto_central_onehot = pd.get_dummies(venues_df[['VenueCategory']], prefix="", prefix_sep="")

# add postal, borough and neighborhood column back to dataframe
toronto_central_onehot['PostalCode'] = venues_df['PostalCode'] 
toronto_central_onehot['Borough'] = venues_df['Borough'] 
toronto_central_onehot['Neighborhoods'] = venues_df['Neighborhood'] 

# move postal, borough and neighborhood column to the first column
fixed_columns = list(toronto_central_onehot.columns[-3:]) + list(toronto_central_onehot.columns[:-3])
toronto_central_onehot = toronto_central_onehot[fixed_columns]

print(toronto_central_onehot.shape)
toronto_central_onehot.head()

(1704, 236)


PostalCode       Borough Neighborhoods  Accessories Store  Adult Boutique  \
0        M4E  East Toronto   The Beaches                  0               0   
1        M4E  East Toronto   The Beaches                  0               0   
2        M4E  East Toronto   The Beaches                  0               0   
3        M4E  East Toronto   The Beaches                  0               0   
4        M4E  East Toronto   The Beaches                  0               0   

   Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                  0        0                   0             0   
1                  0        0                   0             0   
2                  0        0                   0             0   
3                  0        0                   0             0   
4                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Arepa Restaurant  Art Gallery  Art Museum  \
0             0         0                 0            0           0   
1             0         0                 0            0           0   
2             0         0                 0            0           0   
3             0         0                 0            0           0   
4             0         0                 0            0           0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Workshop  \
0                    0                 0                   0              0   
1                    0                 0                   0              0   
2                    0                 0                   0              0   
3                    0                 0                   0              0   
4                    0                 0                   0              0   

   BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Stadium  \
0          0           0           0       0     0    0                 0   
1          0           0           0       0     0    0                 0   
2          0           0           0       0     0    0                 0   
3          0           0           0       0     0    0                 0   
4          0           0           0       0     0    0                 0   

   Basketball Stadium  Beach  Beer Bar  Beer Store  Belgian Restaurant  \
0                   0      0         0           0                   0   
1                   0      0         0           0                   0   
2                   0      0         0           0                   0   
3                   0      0         0           0                   0   
4                   0      0         0           0                   0   

   Bistro  Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  \
0       0              0          0         0                     0   
1       0              0          0         0                     0   
2       0              0          0         0                     0   
3       0              0          0         0                     0   
4       0              0          0         0                     0   

   Breakfast Spot  Brewery  Bubble Tea Shop  Burger Joint  Burrito Place  \
0               0        0                0             0              0   
1               0        0                0             0              0   
2               0        0                0             0              0   
3               0        0                0             0              0   
4               0        0            

**Get the frequency of occurance of each category in an area**

In [27]:
toronto_central_venues_freq = toronto_central_onehot.groupby(['PostalCode', 'Borough', 'Neighborhoods']).mean().reset_index()
print(toronto_central_venues_freq.shape)
toronto_central_venues_freq.head()

(38, 236)


PostalCode          Borough                  Neighborhoods  \
0        M4E     East Toronto                    The Beaches   
1        M4K     East Toronto    The Danforth West,Riverdale   
2        M4L     East Toronto  The Beaches West,India Bazaar   
3        M4M     East Toronto                Studio District   
4        M4N  Central Toronto                  Lawrence Park   

   Accessories Store  Adult Boutique  Afghan Restaurant  Airport  \
0                0.0             0.0                0.0      0.0   
1                0.0             0.0                0.0      0.0   
2                0.0             0.0                0.0      0.0   
3                0.0             0.0                0.0      0.0   
4                0.0             0.0                0.0      0.0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                 0.0           0.0             0.0              0.0   
1                 0.0           0.0             0.0              0.0   
2                 0.0           0.0             0.0              0.0   
3                 0.0           0.0             0.0              0.0   
4                 0.0           0.0             0.0              0.0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  \
0               0.0              0.00000           0.0       0.0   
1               0.0              0.02381           0.0       0.0   
2               0.0              0.00000           0.0       0.0   
3               0.0              0.05000           0.0       0.0   
4               0.0              0.00000           0.0       0.0   

   Arepa Restaurant  Art Gallery  Art Museum  Arts & Crafts Store  \
0               0.0          0.0         0.0                  0.0   
1               0.0          0.0         0.0                  0.0   
2               0.0          0.0         0.0                  0.0   
3               0.0          0.0         0.0                  0.0   
4               0.0          0.0         0.0                  0.0   

   Asian Restaurant  Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  \
0               0.0                 0.0            0.0        0.0         0.0   
1               0.0                 0.0            0.0        0.0         0.0   
2               0.0                 0.0            0.0        0.0         0.0   
3               0.0                 0.0            0.0        0.0         0.0   
4               0.0                 0.0            0.0        0.0         0.0   

   Bagel Shop   Bakery   Bank    Bar  Baseball Stadium  Basketball Stadium  \
0         0.0  0.00000  0.000  0.000               0.0                 0.0   
1         0.0  0.02381  0.000  0.000               0.0                 0.0   
2         0.0  0.00000  0.000  0.000               0.0                 0.0   
3         0.0  0.05000  0.025  0.025               0.0                 0.0   
4         0.0  0.00000  0.000  0.000               0.0                 0.0   

   Beach  Beer Bar  Beer Store  Belgian Restaurant  Bistro  Boat or Ferry  \
0    0.0       0.0         0.0                 0.0     0.0            0.0   
1    0.0       0.0         0.0                 0.0     0.0            0.0   
2    0.0       0.0         0.0                 0.0     0.0            0.0   
3    0.0       0.0         0.0                 0.0     0.0            0.0   
4    0.0       0.0         0.0                 0.0     0.0            0.0   

   Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  Brewery  \
0    0.00000       0.0                   0.0           0.000  0.00000   
1    0.02381       0.0                   0.0           0.000  0.02381   
2    0.00000       0.0                   0.0           0.000  0.05000   
3    0.02500       0.0                   0.0           0.025  0.02500   
4    0.00000       0.0                   0.0           0.000  0.00000   

   Bubble Tea Shop  Burger Joint  Burrito Place  Bus Line  Butcher     Café  \
0          0.0

**Get 10 most occurance venue types in each area**

In [28]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
areaColumns = ['PostalCode', 'Borough', 'Neighborhoods']
freqColumns = []
for ind in np.arange(num_top_venues):
    try:
        freqColumns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        freqColumns.append('{}th Most Common Venue'.format(ind+1))
columns = areaColumns+freqColumns
# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['PostalCode'] = toronto_central_venues_freq['PostalCode']
neighborhoods_venues_sorted['Borough'] = toronto_central_venues_freq['Borough']
neighborhoods_venues_sorted['Neighborhoods'] = toronto_central_venues_freq['Neighborhoods']

for ind in np.arange(toronto_central_venues_freq.shape[0]):
    row_categories = toronto_central_venues_freq.iloc[ind, :].iloc[3:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    neighborhoods_venues_sorted.iloc[ind, 3:] = row_categories_sorted.index.values[0:num_top_venues]

neighborhoods_venues_sorted.sort_values(freqColumns, inplace=True)
neighborhoods_venues_sorted

PostalCode           Borough  \
27        M5V  Downtown Toronto   
31        M6H      West Toronto   
32        M6J      West Toronto   
34        M6P      West Toronto   
35        M6R      West Toronto   
26        M5T  Downtown Toronto   
3         M4M      East Toronto   
25        M5S  Downtown Toronto   
14        M5B  Downtown Toronto   
33        M6K      West Toronto   
21        M5L  Downtown Toronto   
29        M5X  Downtown Toronto   
17        M5G  Downtown Toronto   
15        M5C  Downtown Toronto   
18        M5H  Downtown Toronto   
16        M5E  Downtown Toronto   
19        M5J  Downtown Toronto   
20        M5K  Downtown Toronto   
13        M5A  Downtown Toronto   
9         M4V   Central Toronto   
28        M5W  Downtown Toronto   
11        M4X  Downtown Toronto   
7         M4S   Central Toronto   
4         M4N   Central Toronto   
22        M5N   Central Toronto   
1         M4K      East Toronto   
30        M6G  Downtown Toronto   
8         M4T   Central Toronto   
0         M4E      East Toronto   
12        M4Y  Downtown Toronto   
37        M7Y      East Toronto   
5         M4P   Central Toronto   
10        M4W  Downtown Toronto   
2         M4L      East Toronto   
23        M5P   Central Toronto   
36        M6S      West Toronto   
24        M5R   Central Toronto   
6         M4R   Central Toronto   

                                        Neighborhoods 1st Most Common Venue  \
27  CN Tower,Bathurst Quay,Island airport,Harbourf...        Airport Lounge   
31                        Dovercourt Village,Dufferin                Bakery   
32                            Little Portugal,Trinity                   Bar   
34                       High Park,The Junction South                   Bar   
35                              Parkdale,Roncesvalles        Breakfast Spot   
26            Chinatown,Grange Park,Kensington Market                  Café   
3                                     Studio District                  Café   
25                      Harbord,University of Toronto                  Café   
14                            Ryerson,Garden District        Clothing Store   
33         Brockton,Exhibition Place,Parkdale Village           Coffee Shop   
21                      Commerce Court,Victoria Hotel           Coffee Shop   
29              First Canadian Place,Underground city           Coffee Shop   
17                                 Central Bay Street           Coffee Shop   
15                                     St. James Town           Coffee Shop   
18                             Adelaide,King,Richmond           Coffee Shop   
16                                        Berczy Park           Coffee Shop   
19    Harbourfront East,Toronto Islands,Union Station           Coffee Shop   
20            Design Exchange,Toronto Dominion Centre           Coffee Shop   
13                           Harbourfront,Regent Park           Coffee Shop   
9   Deer Park,Forest Hill SE,Rathnelly,South Hill,...           Coffee Shop   
28                    Stn A PO Boxes 25 The Esplanade           Coffee Shop   
11                         Cabbagetown,St. James Town           Coffee Shop   
7                                          Davisville          Dessert Shop   
4                                       Lawrence Park    Dim Sum Restaurant   
22                                           Roselawn                Garden   
1                         The Danforth West,Riverdale      Greek Restaurant   
30                                           Christie         Grocery Store   
8                          Moore Park,Summerhill East                   Gym   
0                                         The Beaches  Gym / Fitness Center   
12                               Church and Wellesley   Japanese Restaurant   
37   Business reply mail Processing Centre969 Eastern    Light Rail Station   
5                                    Davisville North                  Park   
10                            

**Clustering areas**  
Use KMeans algorigthm, try to cluster the toronto central areas into 3 clusters

In [29]:
kclusters = 3

toronto_central_venues_freq_clustering = toronto_central_venues_freq.drop(['PostalCode', 'Borough', 'Neighborhoods'], 1)

kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_central_venues_freq_clustering)

toronto_central_clustered_df = toronto_central_df
toronto_central_clustered_df['Cluster'] = kmeans.labels_

toronto_central_clustered_df = toronto_central_clustered_df.join(neighborhoods_venues_sorted.drop(['Borough', 'Neighborhoods'], 1).set_index('PostalCode'), on='PostalCode')
toronto_central_clustered_df.sort_values(['Cluster'] + freqColumns, inplace=True)
toronto_central_clustered_df

PostalCode           Borough  \
8         M4T   Central Toronto   
10        M4W  Downtown Toronto   
23        M5P   Central Toronto   
27        M5V  Downtown Toronto   
31        M6H      West Toronto   
32        M6J      West Toronto   
34        M6P      West Toronto   
35        M6R      West Toronto   
26        M5T  Downtown Toronto   
3         M4M      East Toronto   
25        M5S  Downtown Toronto   
14        M5B  Downtown Toronto   
33        M6K      West Toronto   
21        M5L  Downtown Toronto   
29        M5X  Downtown Toronto   
17        M5G  Downtown Toronto   
15        M5C  Downtown Toronto   
18        M5H  Downtown Toronto   
16        M5E  Downtown Toronto   
19        M5J  Downtown Toronto   
20        M5K  Downtown Toronto   
13        M5A  Downtown Toronto   
9         M4V   Central Toronto   
28        M5W  Downtown Toronto   
11        M4X  Downtown Toronto   
7         M4S   Central Toronto   
4         M4N   Central Toronto   
1         M4K      East Toronto   
30        M6G  Downtown Toronto   
0         M4E      East Toronto   
12        M4Y  Downtown Toronto   
37        M7Y      East Toronto   
5         M4P   Central Toronto   
2         M4L      East Toronto   
36        M6S      West Toronto   
24        M5R   Central Toronto   
6         M4R   Central Toronto   
22        M5N   Central Toronto   

                                         Neighborhood   Latitude  Longitude  \
8                          Moore Park,Summerhill East  43.689574 -79.383160   
10                                           Rosedale  43.679563 -79.377529   
23                 Forest Hill North,Forest Hill West  43.696948 -79.411307   
27  CN Tower,Bathurst Quay,Island airport,Harbourf...  43.628947 -79.394420   
31                        Dovercourt Village,Dufferin  43.669005 -79.442259   
32                            Little Portugal,Trinity  43.647927 -79.419750   
34                       High Park,The Junction South  43.661608 -79.464763   
35                              Parkdale,Roncesvalles  43.648960 -79.456325   
26            Chinatown,Grange Park,Kensington Market  43.653206 -79.400049   
3                                     Studio District  43.659526 -79.340923   
25                      Harbord,University of Toronto  43.662696 -79.400049   
14                            Ryerson,Garden District  43.657162 -79.378937   
33         Brockton,Exhibition Place,Parkdale Village  43.636847 -79.428191   
21                      Commerce Court,Victoria Hotel  43.648198 -79.379817   
29              First Canadian Place,Underground city  43.648429 -79.382280   
17                                 Central Bay Street  43.657952 -79.387383   
15                                     St. James Town  43.651494 -79.375418   
18                             Adelaide,King,Richmond  43.650571 -79.384568   
16                                        Berczy Park  43.644771 -79.373306   
19    Harbourfront East,Toronto Islands,Union Station  43.640816 -79.381752   
20            Design Exchange,Toronto Dominion Centre  43.647177 -79.381576   
13                           Harbourfront,Regent Park  43.654260 -79.360636   
9   Deer Park,Forest Hill SE,Rathnelly,South Hill,...  43.686412 -79.400049   
28                    Stn A PO Boxes 25 The Esplanade  43.646435 -79.374846   
11                         Cabbagetown,St. James Town  43.667967 -79.367675   
7                                          Davisville  43.704324 -79.388790   
4                                       Lawrence Park  43.728020 -79.388790   
1                         The Danforth West,Riverdale  43.679557 -79.352188   
30                                           Christie  43.669542 -79.422564   
0                                         The Beaches  43.676357 -79.293031   
12                               Church and Wellesley  43.665860 -79.383160   
37   Business reply mail Processing Centre969 Eastern  43.662744 -79.321558   
5                             

**Show those clusters onto a map**

In [30]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=12)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, post, bor, poi, cluster in zip(toronto_central_clustered_df['Latitude'], toronto_central_clustered_df['Longitude'], toronto_central_clustered_df['PostalCode'], toronto_central_clustered_df['Borough'], toronto_central_clustered_df['Neighborhood'], toronto_central_clustered_df['Cluster']):
    label = folium.Popup('{} ({}): {} - Cluster {}'.format(bor, post, poi, cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

Upon observing the result, we can name the clusters as follow:  
- Cluster 0: Living area (with mostly park, trail, school; and some small businesses)
- Cluster 1: Roselawn - Central Toronto (nothing here except a garden)
- Cluster 2: Business area (with lots of business venues)